### DISCUSSION TITLE
Drawing live line on Canvas

### DISCUSSION THREAD

**Discussion by matthewcrews** [fsprojects/Avalonia.FuncUI#322] [Q&A] [answered]
I have been developing a simple tool for dragging and connecting blocks. I've added the ability to click on an output node (yellow button on the right) and I'm drawing a red line to wherever the cursor is pointing to. The problem is that it only updates to the cursor's position when it's inside another button on the screen.

Here's a snapshot of the line drawing to where the cursor currently is:
![image](https://github.com/fsprojects/Avalonia.FuncUI/assets/1593215/aeeb9fc6-0872-468f-be68-a291690ff600)

If I move the cursor outside of the buttons I get this though:
![image](https://github.com/fsprojects/Avalonia.FuncUI/assets/1593215/170bada0-c87b-4dc1-8194-7cfe760f13be)

Notice how the cursor has moved to the right but the green line is no longer tracking the cursor. I tried adding an event handler for `onPointerMoved` to the Canvas but that doesn't seem to be working:

```fsharp
Canvas.create [
    Canvas.onPointerMoved (fun e ->
        let newPointerLocation = e.GetPosition null
        Msg.Move newPointerLocation |> dispatch)
    Canvas.children [
```

You can see the full code at the link below. To recreate the issue, clone the repo, run the program, click on the right yellow button, and move the cursor around.

Code for Line Drawing:
https://github.com/matthewcrews/AvaloniaDragAndDrop/blob/5c1899517061d35da9ad859fc922e3250871f2d1/Program.fs#L199

**Comment by JaggerJo** [comment]
You need to capture the pointer if you want to get movements even when it leaves the element.

**Reply by JaggerJo** [reply]
https://reference.avaloniaui.net/api/Avalonia.Input/Pointer/2F69A856

**Reply by matthewcrews** [reply]
My apologies, but I'm failing to see how to use the ability to capture the pointer 😔.

**Reply by matthewcrews** [reply]
I've done some searching for tutorials or examples of how to use Pointer Capture, and I haven't found any for AvaloniaUI. It's not clear to me where to put the pointer capture code or how to use it.

**Reply by JaggerJo** [reply]
From on top of my mind:

When entering the line drawing mode (pointer down) you need to capture the pointer. This makes sure you get all the events even if the pointer moves outside the element. When you are done release the pointer again.

I don't have a code sample at hand. Let me know if you are stuck and I'll take a look.

**Reply by matthewcrews** [reply]
I'll be honest. I am thoroughly stuck. I apologize for being dense. I'm trying to thrash around with the API to see if I can even figure out where the `Capture` method is so that I can use it.

I have not been able to find a tutorial or example where someone is doing this, so it's really hard for me to figure out how to use the API. That's more a statement of my ability as a dev than a critique of the API 😊.

Is this something that needs to go in the `onPointerMoved` event handler? I'm just confused about where I need to put this code and how I need to wire it into the Elmish loop.

No hurry but any assistance is greatly appreciated 🙇.

**Reply by matthewcrews** [reply]
I'm guessing I need to perform this capture in the `Button.onPointerPressed` method. Something like this:

```
Button.onPointerPressed (fun e ->
    let pointer = new Input.Pointer(0, Input.PointerType.Mouse, true)
    pointer.Capture (pointer :> Input.IInputElement)
    let newPointerLocation = e.GetPosition null
    Msg.Selection (Selection.Block (blockIdx, newPointerLocation)) |> dispatch)
```

But I'm not confident. The compiler also says the cast from `Pointer` to `IInputElement` is invalid.

**Reply by JaggerJo** [reply]
Yes, exactly. And when you're done / user aborts selection you need to Release the pointer again. 

Does it work?

**Reply by matthewcrews** [reply]
No. The code I wrote above doesn't work because the compiler says that `Pointer` cannot be cast to `IInputElement`. I also cannot find any example of anyone capturing a pointer online. The only thing I find is the API documentation which just tells me the method type signature but no working examples which is why I'm so lost. I also don't see how to "release" a pointer either, so I don't know how to do that. From what I can tell in my search, no one has an example of how to do this.

**Reply by JaggerJo** [reply, marked as answer]
<img width="871" alt="Screenshot 2023-06-29 at 13 47 05" src="https://github.com/fsprojects/Avalonia.FuncUI/assets/13090415/567207cd-22b3-47c2-9aee-fe0ef371e6f0">

**Reply by matthewcrews** [reply]
Okay, so I did something silly and it works now. I added a color to the canvas and now everything works as intended.

```
Canvas.create [
    Canvas.background "DarkSlateGray" // <-- Adding this line makes it work
    // Rest of code here
```

UI looks like this now:
![image](https://github.com/fsprojects/Avalonia.FuncUI/assets/1593215/b800da41-e9b4-455a-80e9-266a5a456fad)

**Reply by JaggerJo** [reply]
Ahh, did not see this. Yeah, you need a non transparent background for it to fire (hit testing, yada yada 😄).

Capturing still is an interesting thing to know about.

**Reply by matthewcrews** [reply]
Agreed, I'm sure it will be useful at some point. And now there's an example of how to do it on the internet!

Also, good to know about needing a non-transparent Canvas for this to work.

**Comment by Numpsy** [comment]
I've never tried doing anything like this myself, but - does https://github.com/wieslawsoltes/NodeEditor do the sort of thing that you're trying?

**Reply by matthewcrews** [reply]
It does. I am trying to make something similar to that. I have a hard time following the code in that repo because it depends on using XAML/C#, which I am not as familiar with.
